### PR TITLE
Allow filtering anchors/delegations and make filtering infallible

### PR DESCRIPTION
**Pull request by thibault-martinez**
- Make all filtering methods infallible
- Add filtering options for anchors and delegations

### CODE REVIEW

**Review by tadzik — DISMISSED** [review]

**Review by kwek20 — DISMISSED** [review]

File: sdk/src/wallet/core/mod.rs [hunk 1]
```diff
@@ -358,50 +379,75 @@ where
                 }
```
Comment (Thoralf-M) [inline sdk/src/wallet/core/mod.rs:384]: Can't comment there, L377 needs additional anchors, delegations .is_none()

Comment (thibault-martinez) [reply sdk/src/wallet/core/mod.rs:384]: Good catch, done

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/wallet/core/mod.rs [hunk 2]
```diff
@@ -353,55 +374,85 @@ where
...
                 // If ids are provided, only return them and no other outputs.
-                if filter.account_ids.is_none() && filter.foundry_ids.is_none() && filter.nft_ids.is_none() {
+                if filter.account_ids.is_none()
+                    && filter.anchor_ids.is_none()
+                    && filter.foundry_ids.is_none()
+                    && filter.nft_ids.is_none()
+                    && filter.delegation_ids.is_none()
+                {
                     filtered_outputs.push(output.clone());
                 }
```
Comment (Alex6323) [inline sdk/src/wallet/core/mod.rs:384]: A bit weird this `if` tbh. Also ... seems like something that can easily introduce a bug. I wonder if there's a better way. Also ... is the comment correct?

Comment (thibault-martinez) [reply sdk/src/wallet/core/mod.rs:384]: Comment looks correct to me yeah

**Review by Alex6323 — DISMISSED** [review]

File: sdk/src/wallet/core/mod.rs (hunk 2 above)
Comment (tadzik) [reply sdk/src/wallet/core/mod.rs:384]: > seems like something that can easily introduce a bug

Yeah, I believe that's what https://github.com/iotaledger/iota-sdk/pull/1568/files#r1382091913 was.

I can't think of a way to make this compile-time-impossible to forget, but we can at least make sure that we have a unit test that takes this path to maximize our chances.

Comment (tadzik) [reply sdk/src/wallet/core/mod.rs:384]: The comment's correct, but it did confuse me too due to describing an `else` of that `if`, if that makes sense. The code itself reads more like something like

`// Always include the output if we're not filtering by IDs`

to me.

**Review by tadzik — DISMISSED** [review]

File: sdk/src/wallet/core/mod.rs (hunk 2 above)
Comment (thibault-martinez) [reply sdk/src/wallet/core/mod.rs:384]: I'll put your comment but without the always as it's a bit too strong, it still depends on the other filtering options

**Review by Alex6323 — APPROVED** [review]

**Review by tadzik — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

**Pull request merged**